wire-ios-sync-engine/Source/Use cases/SearchUsersUseCase.swift

```diff
@@ -0,0 +1,103 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireUtilities
+
+public protocol SearchUsersUseCaseProtocol {
+
+    func invoke(
+        query: String,
+        options: SearchOptions,
+        messageProtocol: MessageProtocol?
+    ) async throws -> SearchResult
+}
+
+public class SearchUsersUseCase: SearchUsersUseCaseProtocol {
+
+    // MARK: - Properties
+
+    private let context: NSManagedObjectContext
+    private let searchDirectory: SearchDirectory?
+    private let isFederationUsageAllowed: Bool
+    private var activeSearchTask: SearchTask?
+
+    deinit {
+        searchDirectory?.tearDown()
+    }
+
+    // MARK: - Initialization
+
+    public init(
+        context: NSManagedObjectContext,
+        searchDirectory: SearchDirectory?,
+        isFederationUsageAllowed: Bool
+    ) {
+        self.context = context
+        self.searchDirectory = searchDirectory
+        self.isFederationUsageAllowed = isFederationUsageAllowed
+    }
+
+    // MARK: - Public Interface
+
+    public func invoke(
+        query: String,
+        options: SearchOptions,
+        messageProtocol: MessageProtocol?
+    ) async throws -> SearchResult {
+        activeSearchTask?.cancel()
+        activeSearchTask = nil
+
+        searchDirectory?.updateIncompleteMetadataIfNeeded()
+
+        let (selfDomain, team) = await context.perform {
+            let selfUser = ZMUser.selfUser(in: self.context)
+            return (selfUser.domain, selfUser.membership?.team)
+        }
+
+        let searchDomain = isOtherDomainSearchAllowed(messageProtocol) ? nil : selfDomain
+        let request = SearchRequest(
+            query: query.trim(),
+            searchDomain: searchDomain,
+            searchOptions: options,
+            team: team
+        )
+
+        return try await withCheckedThrowingContinuation { continuation in
+            let task = searchDirectory?.perform(request)
+            task?.addResultHandler { result, isCompleted in
+                if isCompleted {
```

Comment:
Is this called multiple times? Will it always eventually be completed?

Reply:
Yes, it triggers  several different searches when we start a search task:
        ```
      performRemoteSearchForServices()

        // search People or groups
        performLocalLookup()
        performLocalSearch()

        // v1
        performUserLookup()
        performRemoteSearchForTeamUser()
        // v2+
        performRemoteSearch()
```

We need to update the UI once we have all the responses.